src/extractor/guidance/intersection_handler.cpp

```diff
@@ -383,131 +394,147 @@ std::size_t IntersectionHandler::findObviousTurn(const EdgeID via_edge,
-            best != 0 && obviousByRoadClass(in_data.road_classification,
-                                            current_best_class,
-                                            out_data.road_classification);
+    const auto End = end(out_ways);
+    // dead end, or otherwise no exitable ways
+    if (num_valid_entry == 0)
+        return 0;
 
-        if ((!other_is_better_choice_by_priority && deviation < best_deviation) ||
-            is_better_choice_by_priority)
-        {
-            best_deviation = deviation;
-            best = i;
-        }
+    // sort out ways by what they share with the in way
+    // entry allowed, lowest deviation to greatest deviation, name, classification and priority
+    const auto sort_by = [](const out_way &lhs, const out_way &rhs) {
+        auto left_tie =
+            std::tie(lhs.same_classification, lhs.same_or_higher_priority);
+        auto right_tie =
+            std::tie(rhs.same_classification, rhs.same_or_higher_priority);
+        return lhs.entry_allowed > rhs.entry_allowed ||
+               (lhs.entry_allowed == rhs.entry_allowed &&
+                (lhs.deviation_from_straight < rhs.deviation_from_straight ||
+                 (lhs.deviation_from_straight == rhs.deviation_from_straight &&
+                  left_tie > right_tie)));
+    };
+    std::stable_sort(begin(out_ways), end(out_ways), sort_by);
+
+    // best candidate is the first way sorted to have the smallest deviation from
+    // straight and is same or higher priority class
+    const auto sameOrHigher = [&](const out_way &way) {
+        return way.entry_allowed && way.same_or_higher_priority;
+    };
+
+    const auto notLowPriority = [&](const out_way &way) {
+        return way.entry_allowed &&
+               !node_based_graph.GetEdgeData(way.road->eid)
+                    .road_classification.IsLowPriorityRoadClass();
+    };
+    auto best_candidate_it = std::find_if(begin(out_ways), End, sameOrHigher);
+    if (best_candidate_it == End)
+        best_candidate_it = std::find_if(begin(out_ways), End, notLowPriority);
+
+    if (best_candidate_it != End)
+    {
+        best = best_candidate_it->index;
+        best_deviation = best_candidate_it->deviation_from_straight;
     }
 
-    // We don't consider empty names a valid continue feature. This distinguishes between missing
-    // names and actual continuing roads.
-    if (in_data.name_id == EMPTY_NAMEID)
-        best_continue = 0;
-
+    // No non-low priority roads? Declare no obvious turn
     if (best == 0)
         return 0;
 
+    // get a count of number of ways from that intersection that qualify to have
+    // continue instruction because they share a name with the approaching way
     const std::pair<std::int64_t, std::int64_t> num_continue_names = [&]() {
         std::int64_t count = 0, count_valid = 0;
-        if (in_data.name_id != EMPTY_NAMEID)
+        std::size_t i, last = out_ways.size();
```

Comment:
`last` is somewhat a synonym for the element itself. Can we stick with size. Also don't use `i` as index, ever. Searching in the code for usage will be made impossible.

Also, this could be an std::algorithm `find_if` returning false and keeping count.
```
[&count,&count_valid]( ... way)
{
    if( same_name_id )
       ++count;
    if( entry_allowed)
       ++valid;
    return false;
}
```

Or (since it is small), directly do two algorithms using `count_if` with dedicated lamdas.